src/asset_manager/asset.rs

```diff
@@ -0,0 +1,200 @@
+///
+/// ### The concept
+///
+/// An asset is loaded in two steps:
+///
+/// 1) Load asset data (`AssetData`)
+/// 2) Turn asset data into asset (`from_data`)
+///
+/// # Examples
+///
+/// Here is an example implementation of an
+/// asset:
+///
+/// ```
+/// use std::io::{Read, BufRead, BufReader};
+///
+/// use amethyst::asset_manager::{Asset, AssetFormat, Import, ImportError};
+/// use amethyst::Context;
+///
+/// struct Table {
+///     num_rows: u32,
+///     num_columns: u32,
+///     data: Box<[i32]>,
+/// }
+///
+/// struct WhitespaceTable;
+///
+/// impl Asset for Table {
+///     type Data = Self;
+///     type Error = (); // Should be some InconsistentSizeError (if data.len() != num_rows * num_columns)
+///
+///     fn category() -> &'static str {
+///         // Tables will be in the "tables" folder
+///
+///         "tables"
+///     }
+///
+///     fn from_data(data: Self, _: &mut Context) -> Result<Self, ()> {
+///         Ok(data) // Ommitted here: Check for size
+///     }
+/// }
+///
+/// impl AssetFormat for WhitespaceTable {
+///     fn file_extensions() -> &'static [&'static str] {
+///         const EXTENSIONS: &'static [&'static str] = &["wst"];
+///
+///         EXTENSIONS
+///     }
+/// }
+///
+/// impl Import<<Table as Asset>::Data> for WhitespaceTable {
+///     fn import(&self, bytes: Box<[u8]>) -> Result<Table, ImportError> {
+///         let mut num_columns = 0;
+///         let mut old_columns = 0;
+///         let mut values = Vec::new();
+///         let mut reader = bytes.as_ref();
+///
+///         for line in reader.lines() {
+///             let line = line?;
+///
+///             for word in line.split_whitespace() {
+///                 num_columns += 1;
+///                 values.push(word.parse().map_err(|x| ImportError::FormatError("Invalid integer".to_string()))?)
+///             }
+///             if old_columns != 0 && num_columns != old_columns {
+///                 return Err(ImportError::FormatError(format!("Number of columns not matching (previous line: {}, this line: {})",
+///                                                             old_columns, num_columns)));
+///             }
+///             old_columns = num_columns;
+///             num_columns = 0;
+///         }
+///
+///         if values.len() == 0 {
+///             return Err(ImportError::FormatError("Empty file".to_string()));
+///         }
+///
+///         Ok(Table {
+///             num_columns: old_columns,
+///             num_rows: values.len() as u32 / old_columns,
+///             data: values.into_boxed_slice(),
+///         })
+///     }
+/// }
+/// ```
+pub trait Asset: Sized {
+    /// The data type, an intermediate format.
+    /// This may also be `Self` if this asset does
+    /// not depend on `Context` (if it is just
+    /// plain data).
+    type Data;
+    /// The error type that may be returned if
+    /// `from_data` fails.
+    type Error: Debug;
```

Comment:
This should be [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html)